templates/tfengine/components/devops/main.tf

```diff
@@ -50,56 +52,54 @@ module "project" {
   source  = "terraform-google-modules/project-factory/google"
   version = "~> 10.2.2"
 
-  name            = "{{.project.project_id}}"
+  name            = var.project.project_id
   {{- if eq .parent_type "organization"}}
-  org_id          = "{{.parent_id}}"
+  org_id          = var.parent_id
   {{- else}}
   org_id          = ""
-  folder_id       = "{{.parent_id}}"
+  folder_id       = var.parent_id
   {{- end}}
-  billing_account = "{{.billing_account}}"
-  lien            = {{get . "enable_lien" true}}
+  billing_account = var.billing_account
+  lien            = true
   # Create and keep default service accounts when certain APIs are enabled.
   default_service_account = "keep"
   # Do not create an additional project service account to be used for Compute Engine.
   create_project_sa = false
-  activate_apis = [
-    "cloudbuild.googleapis.com",
-    "cloudidentity.googleapis.com",
-    {{range get . "project.apis" -}}
-    "{{.}}",
-    {{end -}}
-  ]
+  activate_apis = var.project.apis
 }
 
 # Terraform state bucket, hosted in the devops project.
 module "state_bucket" {
 source  = "terraform-google-modules/cloud-storage/google//modules/simple_bucket"
   version = "~> 1.4"
 
-  name       = "{{.state_bucket}}"
+  name       = var.state_bucket
   project_id = module.project.project_id
-  location   = "{{.storage_location}}"
+  location   = var.storage_location
 }
 
-{{- if not (get .project.owners_group "exists" false)}}
+{{- if $missing_project_owners_group}}
 
 # Devops project owners group.
 module "owners_group" {
   source  = "terraform-google-modules/group/google"
   version = "~> 0.1"
 
-  id = "{{.project.owners_group.id}}"
-  customer_id = "{{.project.owners_group.customer_id}}"
-  {{- if has .project.owners_group "display_name"}}
-  display_name = "{{.project.owners_group.display_name}}"
-  {{- else}}
-  display_name = "{{regexReplaceAll "@.*" .project.owners_group.id ""}}"
-  {{- end}}
-  {{hclField .project.owners_group "description" -}}
-  {{hclField .project.owners_group "owners" -}}
-  {{hclField .project.owners_group "managers" -}}
-  {{hclField .project.owners_group "members" -}}
+  id = var.project.owners_group.id
+  customer_id = var.project.owners_group.customer_id
+  display_name = var.project.owners_group.display_name
+  {{- if has .project.owners_group "description"}}
+  description = var.project.owners_group.description
+  {{- end }}
+  {{- if has .project.owners_group "owners"}}
```

Comment:
Is the reason why we need to do these conditions here those warnings? Can you show me one of those as an example?

Reply:
This is not a warning but an error.

If the variable is not present, it throws because `owners_group` doesn't have `owners` attribute, forcing us to put it in `variables.tf`, but if it's there, it throws because it doesn't have an specified value, so we have to add a `[]` default at `terraform.tfvars`. That works, but adds extra overhead. I prefer to hide them all.

An alternative is to use the default at `variables.tf`, but I already mention what happen with complex data structures

Reply:
I understand it needs to be in `variables.tf`. The thing I'm not too sure is the error thrown when it doesn't have a specified value in the `terraform.tfvars` file. Can you show me an error like that?

Also adding defaults in `terraform.tfvars` is actually not bad. You can probably do something like

`{{get .project.owners_group "description" "" -}}` The last field is the default value.

Reply:
When the variable is not specified but is a simple variable, it requests you to write it:

![Screenshot 2021-06-30 1 04 06 PM](https://user-images.githubusercontent.com/33379285/124010616-a6215b80-d9a4-11eb-8bee-18420026cf64.png)
(Note that the error is because I just closed).

The problem comes with nested variables, in that case it just doesn't request, just throws:

![Screenshot 2021-06-30 1 04 20 PM](https://user-images.githubusercontent.com/33379285/124010578-986bd600-d9a4-11eb-95a5-76d9a58567e7.png)

The problem with defaults is not how good it is. Is that it might conflict with constrains like regex. Let's say that a user doesn't specify a `parent_id`, which is not required by schema. It might end up writing a default empty value that will throw an error when doing `terraform plan` because it's considered invalid

Reply:
I see. Got it. Thanks!